src/v0/sources/shopify/util.js

```diff
@@ -111,11 +119,204 @@ const setAnonymousId = (message) => {
+ * which makes it difficult to comapre the json
+ * @param {*} prevLineItems
+ * @param {*} newLineItems
+ * @returns
+ */
+// https://shopify.dev/docs/api/liquid/objects/line_item#line_item-properties
+const isContainingSameLineItems = (prevLineItems, newLineItems) => {
+  let noOfItems = prevLineItems.length;
+  while (noOfItems > 0) {
+    const modifiedPrev = _.cloneDeep(prevLineItems[noOfItems - 1]);
+    const modifiedNew = _.cloneDeep(newLineItems[noOfItems - 1]);
+    if (modifiedPrev?.properties === null) {
+      modifiedPrev.properties = {};
+    }
+    if (modifiedNew?.properties === null) {
+      modifiedNew.properties = {};
+    }
+    if (JSON.stringify(modifiedPrev) !== JSON.stringify(modifiedNew)) {
+      return false;
+    }
+    noOfItems -= 1;
+  }
+  return true;
+};
+
+/**
+ * This function returns true if payloads are same or Timestamp difference is less than `timeDifferenceForCartEvents` seconds otherwise false
+ * @param {*} prevPayload
+ * @param {*} newPayload
+ */
+const isDuplicateCartPayload = (prevPayload, newPayload) => {
+  // The cart?.items.length param is in the case when rudder identifier is stored in the database
+  const prevPayloadLineItemsLength = prevPayload?.line_items ? prevPayload.line_items.length : prevPayload?.items.length
+  const newPayloadLineItemsLength = newPayload.line_items?.length || newPayload.cart?.items.length
+  if (prevPayloadLineItemsLength !== newPayloadLineItemsLength) {
+    return false;
+  }
+  if (
+    Date.parse(newPayload.updated_at) - Date.parse(prevPayload.updated_at) <
+    timeDifferenceForCartEvents ||
+    newPayloadLineItemsLength === 0 ||
+    isContainingSameLineItems(prevPayload.line_items, newPayload.line_items)
+  ) {
+    return true;
+  }
+  return false;
+};
+
+/**
+ * This Function will check for cart duplication events
+ * @param {*} event
+ */
+const checkForValidRecord = async (newCart, metricMetadata) => {
+  const cartToken = newCart.cart_token || newCart.token;
+  const executeStartTime = Date.now();
+  const redisVal = await RedisDB.getVal(`${cartToken}`);
+  stats.timing('redis_get_latency', executeStartTime, {
+    ...metricMetadata,
+  });
+  stats.increment('shopify_redis_get_data', {
+    ...metricMetadata,
+    timestamp: Date.now(),
+  });
+
+  const oldCart = redisVal?.cart;
+  const anonymousId = redisVal?.anonymousId;
+  await anonymousIdCache.get(cartToken, () => anonymousId); // set anonymousId in cache with cartToken
+  if (!oldCart) {
+    // this is the case for events for which we don't have the values store in redis but are valid events. 
+    // This can be removed afterwards as it is more on backward compatibility side
+
+    return true;
+  }
+  if (isDefinedAndNotNull(oldCart) && isDuplicateCartPayload(oldCart, newCart)) {
+    return false;
+  }
+  const setStartTime = Date.now();
+  await RedisDB.setVal(`${cartToken}`, { anonymousId, cart: newCart });
+  stats.timing('redis_set_latency', setStartTime, {
```

Comment:
ideally, this should happen with in  RedisDB.setVal right?

Reply:
Yess but we are passing destination name as metadata. To send this at redis we need to send metadata to redis  as well. Not sure if this would be right. Your thoughts?